packages/next/src/server/after/after-context.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  requestAsyncStorage,
+  type RequestStore,
+} from '../../client/components/request-async-storage.external'
+import { BaseServerSpan } from '../lib/trace/constants'
+import { getTracer } from '../lib/trace/tracer'
+import type { CacheScope } from './react-cache-scope'
+import { ResponseCookies } from '../web/spec-extension/cookies'
+import type { RequestLifecycleOpts } from '../base-server'
+import type { AfterCallback, AfterTask } from './after'
+import { InvariantError } from '../../shared/lib/invariant-error'
+
+export interface AfterContext {
+  run<T>(requestStore: RequestStore, callback: () => T): T
+  after(task: AfterTask): void
+}
+
+export type AfterContextOpts = {
+  waitUntil: RequestLifecycleOpts['waitUntil'] | undefined
+  onClose: RequestLifecycleOpts['onClose'] | undefined
+  cacheScope: CacheScope | undefined
+}
+
+export function createAfterContext(opts: AfterContextOpts): AfterContext {
+  return new AfterContextImpl(opts)
+}
+
+export class AfterContextImpl implements AfterContext {
+  private waitUntil: RequestLifecycleOpts['waitUntil'] | undefined
+  private onClose: RequestLifecycleOpts['onClose'] | undefined
+  private cacheScope: CacheScope | undefined
+
+  private requestStore: RequestStore | undefined
+
+  private afterCallbacks: AfterCallback[] = []
+
+  constructor({ waitUntil, onClose, cacheScope }: AfterContextOpts) {
+    this.waitUntil = waitUntil
+    this.onClose = onClose
+    this.cacheScope = cacheScope
+  }
+
+  public run<T>(requestStore: RequestStore, callback: () => T): T {
+    this.requestStore = requestStore
+    if (this.cacheScope) {
+      return this.cacheScope.run(() => callback())
+    } else {
+      return callback()
+    }
+  }
+
+  public after(task: AfterTask): void {
+    if (isPromise(task)) {
+      task.catch(() => {}) // avoid unhandled rejection crashes
+      if (!this.waitUntil) {
+        errorWaitUntilNotAvailable()
+      }
+      this.waitUntil(task)
+    } else if (typeof task === 'function') {
+      // TODO(after): will this trace correctly?
+      this.addCallback(() =>
+        getTracer().trace(BaseServerSpan.after, () => task())
+      )
+    } else {
+      throw new Error(
+        '`unstable_after()` must receive a promise or a function as its argument'
+      )
+    }
+  }
+
+  private addCallback(callback: AfterCallback) {
+    if (this.afterCallbacks.length === 0) {
+      // if something is wrong, throw synchronously, bubbling up to the `unstable_after` callsite.
+      if (!this.waitUntil) {
+        errorWaitUntilNotAvailable()
+      }
+      if (!this.requestStore) {
+        throw new InvariantError(
+          'Expected `AfterContext.requestStore` to be initialized'
+        )
+      }
+      if (!this.onClose) {
+        throw new Error(
+          '`unstable_after()` received a function, but Next.js will not be able to run it, because `onClose` is not implemented for the current environment.'
```

Comment:
What will users do with this error? Do we need an `error.md` for this case?

Reply:
Hmm. Actually maybe this should be an InvariantError, If this happens it's almost definitely a bug.